### ISSUE TITLE
Issues with new model registration in NEST

### ISSUE THREAD

**Issue by pnbabu** [nest/nest-simulator#3064]
Following #2990, the model registration and the semantics of `ResetKernel()` have changed. After incorporating these changes to NESTML models, the tests fail when there are multiple calls to the `ResetKernel()`. For example,

```python
nest.ResetKernel()
nest.Install("nestmlmodule")

nrn = nest.Create(“neuron1”)
...
nest.Simulate(100.)
...
nest.ResetKernel()  # removes the neuron1 and neuron2 from the list of registered models
# nest.Install("nestmlmodule”)   # Error: Module 'nestmlmodule' is loaded already.
nrn = nest.Create(“neuron2”)  # Neuron cannot be created - Unknown neuron model
```

In the above example `nestmlmodule` is an external module consisting of models `neuron1` and `neuron2`. According to the new semantics of `ResetKernel()`, a call to the function deregisters the models loaded from an external module. Due to this, the second call to `ResetKernel()` de-registers `neuron1` and `neuron2` which were previously installed from the `nest.Install()` call. Hence, the `neuron2` cannot be created and the script fails. 

A second call to `nest.Install()` after the second `ResetKernel()` also fails as the same module cannot be loaded again. This is a known error as the external modules are still wrapped as SLI modules and can't be unloaded/reloaded.

As a result of these issues, the NESTML master breaks with the latest NEST master. Therefore, for the new model registration to work with NESTML, the module unloading/reloading needs to be in place.

**Comment by terhorstd** [comment]
@heplesser, it seems the module is unloaded but the name stays registered. The second `Install()` should not fail.

**Comment by heplesser** [comment]
Yes, this is a problem I had overlooked. I am not sure if this is already solved in the "new pynest" branch so we could take the solution from there. The key point would be to remember which new-style modules we have loaded and call their `init()` method. One challenge is that right now, we only unload neuron and synapse models on `ResetKernel()` but not backends or connection rules, while the module `init()` registers them all, see e.g., https://github.com/nest/nest-extension-module/blob/d3812c192cbfd86e1d777bf65d2c39c77686784c/src/mymodule.cpp#L106-L128
